GW2EIEvtcParser/EncounterLogic/Strikes/EndOfDragons/AetherbladeHideout.cs

```diff
@@ -43,93 +62,485 @@ protected override CombatReplayMap GetCombatMapInternal(ParsedEvtcLog log)
+                    }
+
+                    // Filter the detonations, we use them only for the end time
+                    var filteredDetonations = new List<EffectEvent>();
+                    if (log.CombatData.TryGetEffectEventsByGUID(EffectGUIDs.AetherbladeHideoutPuzzleCirclesDetonation, out IReadOnlyList<EffectEvent> detonations))
+                    {
+                        foreach (EffectEvent effect in detonations)
+                        {
+                            if (filteredDetonations.Where(x => Math.Abs(x.Time - effect.Time) < ServerDelayConstant).FirstOrDefault() == null)
+                            {
+                                filteredDetonations.Add(effect);
+                            }
+                        }
+                    }
+
+                    foreach (AbstractSingleActor bomb in filteredBombs)
+                    {
+                        (long start, long end) lifespanFirstCircle = (0, 0);
+
+                        // In normal mode, the Echo gains Determined but in challenge mode it doesn't, we use the HP updates instead.
+                        Segment last60HpUpdate = target.GetHealthUpdates(log).Where(x => x.Value > 58 && x.Value <= 60).MinBy(x => x.Start);
+                        Segment last20HpUpdate = target.GetHealthUpdates(log).Where(x => x.Value > 18 && x.Value <= 20).MinBy(x => x.Start);
+
+                        if (last60HpUpdate != null && Math.Abs(bomb.FirstAware - last60HpUpdate.Start) < threshold)
+                        {
+                            EffectEvent detonation = filteredDetonations.Where(x => Math.Abs(bomb.LastAware - x.Time) < threshold).FirstOrDefault();
+                            if (detonation != null)
+                            {
+                                lifespanFirstCircle = (last60HpUpdate.Start, Math.Min(bomb.LastAware, detonation.Time));
+                            }
+                            else // Fallback
+                            {
+                                lifespanFirstCircle = (last60HpUpdate.Start, bomb.LastAware);
+                            }
+                        }
+                        else if (last20HpUpdate != null)
+                        {
+                            EffectEvent detonation = filteredDetonations.Where(x => Math.Abs(bomb.LastAware - x.Time) < threshold).FirstOrDefault();
+                            if (detonation != null)
+                            {
+                                lifespanFirstCircle = (last20HpUpdate.Start, Math.Min(bomb.LastAware, detonation.Time));
+                            }
+                            else // Fallback
+                            {
+                                lifespanFirstCircle = (last20HpUpdate.Start, bomb.LastAware);
+                            }
+                        }
+
+                        long duration = lifespanFirstCircle.end - lifespanFirstCircle.start; // time spent for the first circle to do a rotation
```

Comment:
The rotating circles block here has too many things in common with the one in Environment, it'll be good to refactor it.